File: Difficulty: Hard/Longest Prefix Suffix/longest-prefix-suffix.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

// User function template for C++

class Solution {
  public:
    int lps(string str) {
        // Your code goes here
        int n=str.size();
        vector<int> llps(n,0);
        int i=1;
        int len=0;
        
        while(i<n){
            if(str[i]==str[len]){
                len++;
                llps[i]=len;
                i++;
            }
            else{
                if(len!=0){
                    len=llps[len-1];
                }
                else{
                    llps[i]=0;
                    i++;
                }
            }
        }
        
        return llps[n-1];
    }
};



//{ Driver Code Starts.

int main() {

    ios_base::sync_with_stdio(0);
    cin.tie(NULL);
    cout.tie(NULL);

    int t;
    cin >> t;
    while (t--) {
        string str;
        cin >> str;

        Solution ob;

        cout << ob.lps(str) << "\n";
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Hard/Longest Prefix Suffix/longest-prefix-suffix.py
#User function Template for python3

class Solution:
	def longestPrefixSuffix(self, s):
		n = len(s)
        lps = [0] * n
        i = 0
        j = 1
    
        while j < n:
            if s[i] == s[j]:
                lps[j] = i + 1
                i += 1
                j += 1
            else:
                if i == 0:
                    lps[j] = 0
                    j += 1
                else:
                    i = lps[i - 1]
    
        return lps[n - 1]


#{ 
 # Driver Code Starts
# Testing the function
if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        str_input = input()
        ob = Solution()
        print(ob.longestPrefixSuffix(str_input))
        print("~")
# } Driver Code Ends